everestrs/everestrs/src/lib.rs

```diff
@@ -0,0 +1,196 @@
+    }
+
+    fn from_vec(data: Vec<u8>) -> Self {
+        Self { data }
+    }
+}
+
+#[derive(FromArgs, Debug)]
+/// An everest Node.
+struct Args {
+    /// prefix of installation.
+    #[argh(option)]
+    #[allow(unused)]
+    pub prefix: PathBuf,
+
+    /// configuration yml that we are running.
+    #[argh(option)]
+    #[allow(unused)]
+    pub conf: PathBuf,
+
+    /// module name for us.
+    #[argh(option)]
+    pub module: String,
+}
+
+/// Implements the handling of commands & variables, but has no specific information about the
+/// details of the current module, i.e. it deals with JSON blobs and strings as command names. Code
+/// generation is used to build the concrete, strongly typed abstractions that are then used by
+/// final implementors.
+pub trait GenericModule: Sync {
+    /// Handler for the command `name` on `implementation_id` with the given `parameters`. The return value
+    /// will be returned as the result of the call.
+    fn handle_command(
+        &self,
+        implementation_id: &str,
+        name: &str,
+        parameters: HashMap<String, serde_json::Value>,
+    ) -> Result<serde_json::Value>;
+
+    fn on_ready(&self) {}
+}
+
+pub struct Runtime<T: GenericModule> {
+    // There are two subleties here:
+    // 1. We are handing out pointers to `module_impl` to `cpp_module` for callbacks. The pointers
+    //    must must stay valid for as long as `cpp_module` is alive. Hence `module_impl` must never
+    //    move in memory. Rust can model this through the Pin concept which upholds this guarantee.
+    //    We use a Box to put the object on the heap.
+    // 2. For the same reason, `module_impl` should outlive `cpp_module`, hence should be dropped
+    //    after it. Rust drops fields in declaration order, hence `cpp_module` should come before
+    //    `module_impl` in this struct.
+    cpp_module: cxx::UniquePtr<ffi::Module>,
+    module_impl: Pin<Box<T>>,
+}
+
+impl<T: GenericModule> Runtime<T> {
+    fn ptr_to_module_impl(&self) -> *const c_char {
+        &*self.module_impl.as_ref() as *const T as *const c_char
+    }
+
+    // TODO(hrapp): This function could use some error handling.
+    pub fn from_commandline(module_impl: T) -> Self {
+        let args: Args = argh::from_env();
+        let mut cpp_module = ffi::create_module(
+            &args.module,
+            &args.prefix.to_string_lossy(),
+            &args.conf.to_string_lossy(),
+        );
+        let manifest_json = cpp_module.as_mut().unwrap().initialize();
+        let manifest: schema::Manifest = manifest_json.deserialize();
+        let module = Self {
+            cpp_module,
+            module_impl: Box::pin(module_impl),
+        };
+
+        // Implement all commands for all of our implementations, dispatch everything to the
+        // GenericModule.
+        for (implementation_id, provides) in manifest.provides {
+            let interface_s = module.cpp_module.get_interface(&provides.interface);
```

Comment:
This kind of information encapsulation is a bit problematic.  Unfortunately, in everest-framework, we still keep the structures about which module connects to which and how it is configured as json, instead of proper typed models.
The language bindings, ideally shouldn't care about the full json schema of the manifest / interfaces etc.  They really only should care about the necessary stuff, like their connections and and config options etc.
Because otherwise, like in this case, the binding depends on implementation details (like how the schema looks like etc).

Reply:
I think this is a misread:

It is true that we pass the full schema as a serialized String - which is both wasteful and overkill - but we only deserialize from it what we need from it. For this we need to know the Schema anyways since we only have json in C++ land either. So either in Rust or in the C++ glue code in `everestrs_sys.cpp` we need to use the schema. I choose to do it in Rust and automatically deserialize it with a Framework because this is the safest and least error prone - and also very performant. Should the schema ever change, this problem will also much easier to find and fix than for example in the Python code, where the schema is used implicitly through dictionaries.